### PR TITLE
Convert coding best practices into a markdown document #4

### LINKED ISSUE #4: Convert coding best practices into a markdown document
Currently it refers to a pdf file in the module web site

### PR DESCRIPTION
**Pull request by thenaesh**
Fixes #4
Ready for review.

I have done my best to preserve all the content, though the format is slightly different from the original PDF. A GitHub-rendered copy of the Markdown is available [here](https://github.com/thenaesh/addressbook-level1/blob/4-coding-best-practices-markdown/doc/CodingBestPractices.md).

In particular, code blocks are used where tables containing code were used in the original PDF. I used Markdown blockquotes for the quotes (e.g. the one by Dijkstra) in the original PDF, but I'm not sure if it's the best way to go about doing it as it looks a little weird (I can't find any better way to do it, though).

### CODE REVIEW

File: doc/CodingBestPractices.md [hunk 1]
```diff
@@ -0,0 +1,281 @@
+# Code Quality Guidelines
```
Comment (damithc) [inline doc/CodingBestPractices.md]: Rephrase to `Code Quality Best Practices` and rename the file to match?

File: doc/CodingBestPractices.md [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+* Whenever an arithmetic calculation is performed, check for overflow and 'divide by zero' errors.
+
+### Minimize global variables
+Global variables may be the most convenient way to pass information around, but they do create implicit links between code segments that use the global variable. Avoid them as much as possible.
+
+### Avoid magic numbers
+Avoid indiscriminate use of numbers as constants (e.g. 3.1415 as the mathematical constant PI) all over the code. Define them as named constants, preferably in a central location. A similar logic applies to string literals with special meanings (e.g. "Error 1432").
+```java
+return 3.1415; // BAD
+return PI;      // BETTER
```
Comment (damithc) [inline doc/CodingBestPractices.md]: Align `//`

File: doc/CodingBestPractices.md [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+> Debugging is twice as hard as writing the code in the first place. Therefore, if you write the code as cleverly as possible, you are, by definition, not smart enough to debug it. _--Brian W. Kernighan_
+
+> Premature optimization is the root of all evil in programming. _--Donald Knuth_
+
+### Code for humans
+Always assume that "anyone who reads the code you write is dumber than you (duh)". Make your code understandable to a "dumb" person. The smarter you think you are compared to your teammates, the more effort you need to make your code understandable to them. Even if we do not intend to pass the code to someone else, code quality is still important because we all become 'strangers' to our own code someday. Here are some tips:
+
+* Avoid long methods. Be wary when a method is longer than the computer screen, and take corrective action when it goes beyond 50 LOC (lines of code). The bigger the haystack, the harder it is to find a needle.
+* Limit the depth of nesting. According to the Linux 1.3.53 CodingStyle documentation, "_if you need more than 3 levels of indentation, you're screwed anyway, and should fix your program_". In particular, __avoid arrowhead-style code__.
+
```
Comment (damithc) [inline doc/CodingBestPractices.md]: Give a link to 'arrowhead style'. e.g. https://blog.codinghorror.com/flattening-arrow-code/

File: doc/CodingBestPractices.md [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+```
+```java
+// BETTER
+readData();
+processData();
+displayResult();
+```
+
+#### Sidenote: Abstraction
+> Most programs are written to solve complex problems involving large amounts of intricate details. It is impossible to deal with all these details at the same time. The guiding principle of abstraction stipulates that we capture only details that are relevant to the current perspective or the task at hand. For example, within a certain software component, we might deal with a ‘user’ data type, while ignoring the details contained in the user data item. These details have been ‘abstracted away’ as they do not affect the task of that software component. This is called data abstraction. On the other hand, control abstraction abstracts away details of the actual control flow to focus on tasks at a simplified level. For example, print(“Hello”) is an abstraction of the actual output mechanism within the computer.
```
Comment (damithc) [inline doc/CodingBestPractices.md]: See if you can put the sidenote title inside the quote block

File: doc/CodingBestPractices.md [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+We all get into situations where we need to write some code momentarily that will be disposed of or replaced with better code later. Such code should be marked in some obvious way (e.g. using an embarrassing print message) so that you will not forget to remove them later. It is irresponsible to retain such code in a product release. Production code should be worthy of the quality standards you live by. If a code is important enough to be released, it is important enough to be of production quality.
+
+> Programs should be written and polished until they acquire publication quality. _--Niklaus Wirth_
+
+### Keep looking for ways to improve code quality
+There are many more good coding practices that are not listed above. If you are serious about improving your code quality, try to look out for more such books. Clean code (by Robert C. Martin) and Code complete (by Steve McConnell) are particularly good resources.
+
+## Appendix
+
+### Clean Code vs Good Code
```
Comment (damithc) [inline doc/CodingBestPractices.md]: Make this section a quote block rather than an appendix?

**Comment by thenaesh** [comment]
@damithc I've made the changes requested. Also, a link to the old PDF appeared in README.md in the meantime, which I've updated to point to CodeQualityBestPractices.md as well.

**Comment by damithc** [comment]
Thanks for the contribution @thenaesh 
You may also create a PR to add yourself to the list of contributors.

**Pull request merged**